### ISSUE TITLE
Infinite dependency loop for azure-core

### ISSUE THREAD

**Issue by davidmweber** [python-poetry/poetry#4054]
- **OS version and name**: Ubuntu Linux 20.04 and OSX Catalina
- **Poetry version**: 1.16 (as installed today with poetry self update)
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: [pyproject.toml](https://gist.github.com/davidmweber/6962aba28258e088d327e46347c7b9c8)

## Issue
When running `poetry update` on this file, it gets stuck in an infinite loop. The output of `poetry update -vvv` is 

```text
   1: fact: isodate (0.6.0) depends on six (*)
   1: selecting isodate (0.6.0)
PyPI: 10 packages found for requests-oauthlib >=0.5.0
   1: fact: requests-oauthlib (1.3.0) depends on oauthlib (>=3.0.0)
   1: fact: requests-oauthlib (1.3.0) depends on requests (>=2.0.0)
   1: selecting requests-oauthlib (1.3.0)
   1: derived: oauthlib (>=3.0.0)
PyPI: 20 packages found for pycparser *
   1: selecting pycparser (2.20)
PyPI: 4 packages found for oauthlib >=3.0.0
   1: selecting oauthlib (3.1.0)
   1: fact: azure-core (1.13.0) depends on requests (>=2.18.4)
   1: fact: azure-core (1.13.0) depends on six (>=1.11.0)
   1: selecting azure-core (1.13.0)
   1: fact: azure-core (1.13.0) depends on requests (>=2.18.4)
   1: fact: azure-core (1.13.0) depends on six (>=1.11.0)
   1: selecting azure-core (1.13.0)
   1: fact: azure-core (1.13.0) depends on requests (>=2.18.4)
   1: fact: azure-core (1.13.0) depends on six (>=1.11.0)
   1: selecting azure-core (1.13.0)
   1: fact: azure-core (1.13.0) depends on requests (>=2.18.4)
   1: fact: azure-core (1.13.0) depends on six (>=1.11.0)
   1: selecting azure-core (1.13.0)
   1: fact: azure-core (1.13.0) depends on requests (>=2.18.4)
```
These lines just repeat. 

Here is the full output up to the point where the loop is established:

[loop.txt](https://github.com/python-poetry/poetry/files/6451553/loop.txt)

**Comment by davidmweber** [comment]
The problem appears to be with the azure-iot-hub v 2.4.0. If I pin that to 2.3.0 then the loop disappears. Hope this helps.

**Comment by cpbotha** [comment]
Debugged this together with @davidmweber - it indeed seems to be version 2.4.0 of azure-iot-hub.

You can reproduce by running `poetry update -vvv` with the following minimal pyproject.toml:
https://gist.github.com/cpbotha/530961e8f959dab545c42be2615e12dc

Something awry with that package, but is there some way for poetry not to land in an infinite loop like this?

**Comment by davidmweber** [comment]
Oh but that is pretty.....

**Comment by davidmweber** [comment]
I can confirm this bug still exists for poetry 1.1.7

**Comment by davidmweber** [comment]
I can confirm that this is still present in 1.1.8

**Comment by dimbleby** [comment]
on the master branch, poetry says:
```
Could not parse version constraint: !=3.3*
```
which isn't super-helpful about where that error comes from: but it's a start, and it's not an infinite loop.

The error actually comes from the python version requirements of `azure-iot-hub`, presumably this is a typo for `3.3.*`

**Comment by dimbleby** [comment]
Anyone who cares to get this fixed should submit an MR to that project, fixing this line: https://github.com/Azure/azure-iot-sdk-python/blob/c56e40d82e97f0a4204efc2d7ad260933260245e/azure-iot-hub/setup.py#L83

**Comment by mkniewallner** [comment]
On Poetry `1.1.14`, I was able to install the dependencies just fine (unless explicitly pinning `azure-iot-hub` to `2.4.0` or `2.5.0`, then I would hit the infinite loop, so I believe that something was fixed in `2.6.0`).

On Poetry `master`, as said above, we hit `Could not parse version constraint: !=3.3*` since parsing `python_requires` is possibly stricter in `1.2.x` (for valid reasons).

Since the issue is with the dependency itself now, closing this, since Poetry is not at cause.
